x11docker: init at 5.4.1

jD91mZM2 opened the pull request:
###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS) (I think, it says this is the default)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

This adds a recipe for x11docker, and also the nxagent backend because it supports both seamless and desktop mode.

jD91mZM2 committed:
x11docker: init at 5.4.1

averelld commented:
Just nuking nxproxy would break x2goclient. Ideally we could get multiple outputs from this build, similar to https://git.archlinux.org/svntogit/packages.git/tree/trunk/PKGBUILD?h=packages/nx

jD91mZM2 commented:
IIRC, there *is* a separation of dev/bin outputs possible, maybe I can use it to further difference outputs of nxagent/nxproxy/libs? I'll check it out later

symphorien commented on pkgs/applications/virtualization/x11docker/default.nix:
```diff
@@ -0,0 +1,31 @@
+  nativeBuildInputs = [ makeWrapper ];
+  buildInputs = [ nx-libs xorg.xhost xorg.xinit ];
+
+  phases = [ "unpackPhase" "installPhase" ];
```
prefer `dontBuild = true;` because it will enable all sort of niceties which come by default with the fixupPhase

Ma27 commented:
You also need to patch shebangs, it currently fails with the following error locally:

```
./result/bin/x11docker-gui: /nix/store/xlai4ls0n42jml470pkrxp7cmr56rj4i-x11docker-5.4.1/bin/.x11docker-gui-wrapped: /bin/bash: bad interpreter: No such file or directory
./result/bin/x11docker-gui: line 3: /nix/store/xlai4ls0n42jml470pkrxp7cmr56rj4i-x11docker-5.4.1/bin/.x11docker-gui-wrapped: Success
```

symphorien commented:
> You also need to patch shebangs

this will come for free by not using `phases = ` which disables the default fixupPhase.

jD91mZM2 commented:
Right, I forgot to check the gui. Thanks for the `dontBuild` tip!

jD91mZM2 commented:
The gui fails with

<details>
<summary>Output of `x11docker-gui`</summary>

```
Usage: grep [OPTION]... PATTERN [FILE]...
Try 'grep --help' for more information.
/nix/store/q3ka7ydgcx2ngwyhxf705a3c9nm38jq5-x11docker-5.4.1/bin/.x11docker-gui-wrapped: line 119: command: --: invalid option
command: usage: command [-pVv] command [arg ...]
sed: can't read /usr/share/X11/xkb/rules/base.lst: No such file or directory
x11docker-gui: Did not find executeable 'kaptain'.
If your distribution does not provide package kaptain (>=0.73), look at
kaptain repository:   https://github.com/mviereck/kaptain
Fallback: Will try to use image x11docker/kaptain.

/nix/store/q3ka7ydgcx2ngwyhxf705a3c9nm38jq5-x11docker-5.4.1/bin/.x11docker-gui-wrapped: line 837: export: `Options:': not a valid identifier
/nix/store/q3ka7ydgcx2ngwyhxf705a3c9nm38jq5-x11docker-5.4.1/bin/.x11docker-gui-wrapped: line 839: /share/stdin: No such file or directory
```

</details>

It appears we need to package kaptain for that gui, so why not just disable that for now? The command line client is (at least to me) the most interesting part of x11docker.

Ma27 commented:
I'll see if I have enough time to test the package today. If the `x11docker` tool works fine, I'd say that we could merge this (but we shouldn't move the GUI part into the store path and add a `TODO` comment instead).

jD91mZM2 commented on pkgs/applications/virtualization/x11docker/default.nix:
```diff
@@ -0,0 +1,32 @@
+    #install -D x11docker-gui "$out/bin/x11docker-gui";
+    wrapProgram "$out/bin/x11docker" --prefix PATH : "${PATH_PREFIX}"
+    #wrapProgram "$out/bin/x11docker-gui" --prefix PATH : "${PATH_PREFIX}"
+    # GUI disabled because of missing `kaptain` dependency
```
@Ma27 I think this is what you wanted right?

Ma27 replied:
yes :+1:

mviereck commented:
I am the developer of x11docker, and I stumbled over this thread yet.
Thank you, I am honoured that you like to integrate x11docker into NixOS!

If there is anything needed to adjust x11docker for NixOS, please tell me. I once did some basic test runs in a NixOS VM, but did not checked out all possible setups.

> It appears we need to package kaptain for that gui, so why not just disable that for now? The command line client is (at least to me) the most interesting part of x11docker.

I am currently looking at the `x11docker-gui` issues mentioned above. However, the interesting part is the `x11docker` script itself, and it needs no other files.

x11docker-gui runs as well with image `x11docker/kaptain`. I have the impression that it is used more often than a native `kaptain` installation.

jD91mZM2 commented:
The docker image didn't seem to work, or at least that's what I assume the error is. Cool thing you stumbled upon this so quickly, I was going to email you once this was merged to let you know your program got packaged. It's really awesome and I'm excited to start packaging some proprietary X11 apps :)

Ma27 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -4517,7 +4519,11 @@ in
   nylon = callPackage ../tools/networking/nylon { };
 
-  nxproxy = callPackage ../tools/admin/nxproxy { };
+  nxproxy = throw ''
```
are there any major differences between this `nxproxy` and the `nxproxy` from `nx-libs`? Unless I'd add an alias to `pkgs/top-level/aliases.nix` and probably throw a deprecation warning.

jD91mZM2 replied:
Good idea! How does one throw in a deprecation warning?

Ma27 replied:
I'd recommend to do something like this:

``` nix
{
  # pkgs/top-level/aliases.nix
  nxproxy = lib.warn "Deprecated. <explanation>" nx-libs; # yyyy-mm-dd
}
```

Ma27 commented:
> If there is anything needed to adjust x11docker for NixOS, please tell me. I once did some basic test runs in a NixOS VM, but did not check out all possible setups.

I didn't read through the entire `x11docker-gui` script, but I guess that most of the problems occur since NixOS isn't fully FHS-compliant and e.g. doesn't have a standard `/usr` directory (except for `/usr/bin/env`) because of its special behavior.

Most of the occurrences can be solved by a simple `substitute(InPlace)`. I don't know though, how much effort it'll take to package `kaptain`.

Ma27 commented on pkgs/tools/X11/nx-libs/default.nix:
```diff
@@ -0,0 +1,44 @@
+  enableParallelBuilding = true;
+
+  postPatch = ''
+    find . -type f -executable -exec sed -i 's|^#!/bin/bash$|#!${bash}/bin/bash|g' {} \;
```
This can be achieved with the `patchShebangs` script which should be done for scripts in the store path in the fixup phase automatically.

Ma27 committed:
nx-libs: minor fixes

* Add an alias with a deprecation warning for `nxproxy` to avoid an
  immediate breaking change.

* Use the default shell used in the build environment (`stdenv.shell`)
  for patching. This shell is in the environment and thus used to patch
  scripts using `patchShebangs`. The shell is referenced as `stdenv.shell`
  in Makefiles to patch the remaining occurrences of `/bin/bash` in the
  build environment.

Ma27 commented:
I pushed a simple commit to your branch that fixes my review. If it's fine for you, I'd go ahead and merge as I just tested `x11docker` and it seems to work fine :+1:

jD91mZM2 commented on pkgs/tools/X11/nx-libs/default.nix:
```diff
@@ -20,8 +20,8 @@ stdenv.mkDerivation rec {
-    find . -type f -executable -exec sed -i 's|^#!/bin/bash$|#!${bash}/bin/bash|g' {} \;
-    find . -type f -name Makefile -exec sed -i 's|^\(SHELL:=\)/bin/bash$|\1${bash}/bin/bash|g' {} \;
+    patchShebangs .
+    find . -type f -name Makefile -exec sed -i 's|^\(SHELL:=\)/bin/bash$|\1${stdenv.shell}|g' {} \;
```
Thanks for the commit! Will this line work even if you use a non-compatible shell, such as fish though?

jD91mZM2 commented:
Ignore my review (which I deleted), I didn't realize `stdenv.shell` was always bash. Looks awesome, thanks for the commit! I'm very fine with merging this right now :smile:

Ma27 commented:
@jD91mZM2 thanks!

vcunat commented:
Cross-ref: 10f37a312dc.

Ma27 commented:
@vcunat thanks a lot for fixing this and the elaboration in your commit message. I guess I was wrong here, so I'm sorry!

Just out of interest, are there any proposals on how to properly deprecate attribute names (like a package in this case?)

vcunat commented:
The most recent thread I know: https://github.com/NixOS/rfcs/pull/33
